sdk/python/toolchain/poetry.go

```diff
@@ -0,0 +1,139 @@
+package toolchain
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/logging"
+)
+
+type poetry struct {
+	// The executable path for poetry.
+	poetryExecutable string
+	// The directory that contains the poetry project.
+	directory string
+}
+
+var _ Toolchain = &poetry{}
+
+func newPoetry(directory string) (*poetry, error) {
+	poetryPath, err := exec.LookPath("poetry")
+	if err != nil {
+		return nil, fmt.Errorf("poetry not found on path: %w", err)
+	}
+	logging.V(9).Infof("Python toolchain: using poetry at %s in %s", poetryPath, directory)
+	return &poetry{
+		poetryExecutable: poetryPath,
+		directory:        directory,
+	}, nil
+}
+
+func (p *poetry) InstallDependencies(ctx context.Context,
+	root string, showOutput bool, infoWriter, errorWriter io.Writer,
+) error {
+	poetryCmd := exec.Command(p.poetryExecutable, "install", "--no-ansi") //nolint:gosec
+	poetryCmd.Dir = p.directory
+	poetryCmd.Stdout = infoWriter
+	poetryCmd.Stderr = errorWriter
+	return poetryCmd.Run()
+}
+
+func (p *poetry) ListPackages(ctx context.Context, transitive bool) ([]PythonPackage, error) {
+	args := []string{"list", "-v", "--format", "json"}
+	if !transitive {
+		args = append(args, "--not-required")
+	}
+
+	cmd, err := p.ModuleCommand(ctx, "pip", args...)
```

Comment:
If we are using the poetry toolchain, are we guaranteed to have `pip` installed?

Reply:
Yes, pip is always installed in virtual environments (and any global python installation). https://peps.python.org/pep-0453/